src/nanoemoji/bitmap_tables.py

```diff
@@ -49,54 +53,85 @@
 _CBDT_SMALL_METRIC_PNG_HEADER_SIZE = 5 + 4
 
 
+def _nudge_into_range(arange: range, value: int, max_move: int = 1) -> int:
+    if value in arange:
+        return value
+    if value > max(arange) and value - max_move <= max(arange):
+        return max(arange)
+    if value < min(arange) and value + max_move >= min(arange):
+        return min(arange)
+    return value
+
+
 class BitmapMetrics(NamedTuple):
     x_offset: int
     y_offset: int
     line_height: int
     line_ascent: int
 
     @classmethod
-    def create(cls, config: FontConfig, ppem: int) -> "BitmapMetrics":
+    def create(cls, config: FontConfig, image_data: PNG, ppem: int) -> "BitmapMetrics":
         # https://github.com/googlefonts/noto-emoji/blob/9a5261d871451f9b5183c93483cbd68ed916b1e9/third_party/color_emoji/emoji_builder.py#L109
         ascent = config.ascender
         descent = -config.descender
 
         line_height = round((ascent + descent) * ppem / float(config.upem))
         line_ascent = ascent * ppem / float(config.upem)
 
+        # center within advance
         metrics = BitmapMetrics(
-            x_offset=max(
-                round((_width_in_pixels(config) - config.bitmap_resolution) / 2), 0
+            x_offset=_nudge_into_range(
+                _INT8_RANGE,
+                max(
+                    round(
+                        (
+                            _width_in_pixels(config, image_data)
+                            - config.bitmap_resolution
+                        )
+                        / 2
+                    ),
+                    0,
+                ),
             ),
-            y_offset=round(
-                line_ascent - 0.5 * (line_height - config.bitmap_resolution)
+            y_offset=_nudge_into_range(
+                _INT8_RANGE,
+                round(line_ascent - 0.5 * (line_height - config.bitmap_resolution)),
             ),
             line_height=line_height,
             line_ascent=round(line_ascent),
         )
 
+        # The FontTools errors when values are out of bounds are a bit nasty
+        # so check here for earlier and more helpful termination
+        assert (
+            config.bitmap_resolution in _UINT8_RANGE
+        ), f"bitmap_resolution out of bounds: {config.bitmap_resolution}"
+        assert metrics.y_offset in _INT8_RANGE, f"y_offset out of bounds: {metrics}"
+
         return metrics
 
 
-def _width_in_pixels(config: FontConfig) -> int:
-    return round(
-        config.bitmap_resolution * config.width / (config.ascender - config.descender)
-    )
+def _pixels_to_funits(config: FontConfig, bitmap_pixel_height: int) -> Tuple[int, int]:
+    # the bitmap is vertically scaled to fill the space desc to asc
+    # this gives us a ratio between pixels and upem
+    funits = config.ascender - config.descender
+    return (bitmap_pixel_height, funits)
```

Comment:
the function is named `_pixels_to_funits`, so i'd expect it to actually return the ratio `pixels/funits` (flaot) instead of passing them through as a tuple. It's not doing much besides subtracting ascender-descender

Reply:
Maybe I should make a type. I'm under the impression that by delaying the computation and always doing multiple (enbiggen) then divide we minimize float damage.